### PR TITLE
Save detected l10n of browser on login

### LINKED ISSUE #373: New user with auto detected German language receives activity notification emails in English
### Steps to reproduce
1. Install ownCloud
2. Configure email settings & syscron
3. Create a new user & log into it
4. Set your notification settings (everything on an hourly basis to your own mail address)
5. Check the preferences that are stored in the database (see attached file db_1.txt)
6. Create changes to the files stored in the users ownCloud
7. Receive the notification mail (see attached file mail_1.txt)
8. Manually set the language setting to English and then back to German
9. Verify the settings stored in the database (see attached file db_2.txt)
10. Create changes to the files stored in the users ownCloud
11. Receive the notification mail (see attached file mail_2.txt)

### Expected behaviour

New user with auto detected German language should receive activity notification emails in German

### Actual behaviour

The mails are sent in English until the user manually sets their language to German. Then the mails are sent in a partially translated German / English mix. I believe this happens because the language setting is not actually stored in database when the user first logs in. ownCloud automatically recognizes the users language, but does not save this setting for further use.

After the user has manually set their language preference it is stored in the database. After the setting has been stored the mails start to be translated.

### Attachments

**db_1.txt**
```
MariaDB [oc811]> SELECT * FROM oc_preferences;
+--------+----------------+-------------------------------+------------------------+
| userid | appid          | configkey                     | configvalue            |
+--------+----------------+-------------------------------+------------------------+
| oc     | firstrunwizard | show                          | 0                      |
| oc     | login          | lastLogin                     | 1439361775             |
| oc     | settings       | email                         | user@domain.com        |
| test   | activity       | notify_email_file_changed     | 1                      |
| test   | activity       | notify_email_file_created     | 1                      |
| test   | activity       | notify_email_file_deleted     | 1                      |
| test   | activity       | notify_email_file_restored    | 1                      |
| test   | activity       | notify_email_files_favorites  | 1                      |
| test   | activity       | notify_email_public_links     | 1                      |
| test   | activity       | notify_email_remote_share     | 1                      |
| test   | activity       | notify_email_shared           | 1                      |
| test   | activity       | notify_setting_batchtime      | 3600                   |
| test   | activity       | notify_setting_self           | 1                      |
| test   | activity       | notify_setting_selfemail      | 1                      |
| test   | activity       | notify_stream_file_changed    | 1                      |
| test   | activity       | notify_stream_file_created    | 1                      |
| test   | activity       | notify_stream_file_deleted    | 1                      |
| test   | activity       | notify_stream_file_restored   | 1                      |
| test   | activity       | notify_stream_files_favorites |                        |
| test   | activity       | notify_stream_public_links    | 1                      |
| test   | activity       | notify_stream_remote_share    | 1                      |
| test   | activity       | notify_stream_shared          | 1                      |
| test   | core           | timezone                      | Europe/Berlin          |
| test   | firstrunwizard | show                          | 0                      |
| test   | login          | lastLogin                     | 1439361988             |
| test   | settings       | email                         | user@domain.com        |
+--------+----------------+-------------------------------+------------------------+
26 rows in set (0.00 sec)
```

**mail_1.txt**
```
Hello test,

You are receiving this email because in the last day the following things happened at http://localhost:8080/owncloud-ent-8.1.1/

* You created Test - Today at 08:23:19 AM
* You created Loeschen.txt - Today at 08:23:23 AM
* You changed Loeschen.txt - Today at 08:23:25 AM
* You created Edit.txt - Today at 08:23:30 AM
* You changed Edit.txt - Today at 08:23:33 AM
* You deleted Loeschen.txt - Today at 08:23:36 AM
* You shared ownCloudUserManual.pdf via link - Today at 08:24:17 AM
```

**db_2.txt**
```
MariaDB [oc811]> SELECT * FROM oc_preferences;
+--------+----------------+-------------------------------+--------------------------+
| userid | appid          | configkey                     | configvalue              |
+--------+----------------+-------------------------------+--------------------------+
| oc     | activity       | notify_email_file_changed     | 1                        |
| oc     | activity       | notify_email_file_created     | 1                        |
| oc     | activity       | notify_email_file_deleted     | 1                        |
| oc     | activity       | notify_email_file_restored    | 1                        |
| oc     | activity       | notify_email_files_favorites  | 1                        |
| oc     | activity       | notify_email_public_links     | 1                        |
| oc     | activity       | notify_email_remote_share     | 1                        |
| oc     | activity       | notify_email_shared           | 1                        |
| oc     | activity       | notify_setting_batchtime      | 3600                     |
| oc     | activity       | notify_setting_self           | 1                        |
| oc     | activity       | notify_setting_selfemail      | 1                        |
| oc     | activity       | notify_stream_file_changed    | 1                        |
| oc     | activity       | notify_stream_file_created    | 1                        |
| oc     | activity       | notify_stream_file_deleted    | 1                        |
| oc     | activity       | notify_stream_file_restored   | 1                        |
| oc     | activity       | notify_stream_files_favorites |                          |
| oc     | activity       | notify_stream_public_links    | 1                        |
| oc     | activity       | notify_stream_remote_share    | 1                        |
| oc     | activity       | notify_stream_shared          | 1                        |
| oc     | core           | timezone                      | Europe/Berlin            |
| oc     | firstrunwizard | show                          | 0                        |
| oc     | login          | lastLogin                     | 1439457757               |
| oc     | settings       | email                         | user@domain.com          |
| test   | activity       | notify_email_file_changed     | 1                        |
| test   | activity       | notify_email_file_created     | 1                        |
| test   | activity       | notify_email_file_deleted     | 1                        |
| test   | activity       | notify_email_file_restored    | 1                        |
| test   | activity       | notify_email_files_favorites  | 1                        |
| test   | activity       | notify_email_public_links     | 1                        |
| test   | activity       | notify_email_remote_share     | 1                        |
| test   | activity       | notify_email_shared           | 1                        |
| test   | activity       | notify_setting_batchtime      | 3600                     |
| test   | activity       | notify_setting_self           | 1                        |
| test   | activity       | notify_setting_selfemail      | 1                        |
| test   | activity       | notify_stream_file_changed    | 1                        |
| test   | activity       | notify_stream_file_created    | 1                        |
| test   | activity       | notify_stream_file_deleted    | 1                        |
| test   | activity       | notify_stream_file_restored   | 1                        |
| test   | activity       | notify_stream_files_favorites |                          |
| test   | activity       | notify_stream_public_links    | 1                        |
| test   | activity       | notify_stream_remote_share    | 1                        |
| test   | activity       | notify_stream_shared          | 1                        |
| test   | core           | lang                          | de                       |
| test   | core           | timezone                      | Europe/Berlin            |
| test   | firstrunwizard | show                          | 0                        |
| test   | login          | lastLogin                     | 1439457202               |
| test   | owncloud       | lostpassword                  | XXXXXXXXXXXX             |
| test   | settings       | email                         | user@domain.com          |
+--------+----------------+-------------------------------+--------------------------+
48 rows in set (0.00 sec)
```

**mail_2.txt**
```
Hallo test,

Du erhÃ¤ltst diese E-Mail, da am letzten Tag folgende Dinge auf http://localhost:8080/owncloud-ent-8.1.1/ geschehen sind

* You created Loeschen.txt - Heute um 10:27:02
* You changed Loeschen.txt - Heute um 10:27:04
* You deleted Loeschen.txt - Heute um 10:27:08
* You changed Edit.txt - Heute um 10:27:11
* Du hast ownCloudUserManual.pdf Ã¼ber einen Link freigegeben - Heute um 10:27:19
```

### Server configuration
**ownCloud version:**
oC 8.1.1 EE

**Updated from an older ownCloud or fresh install:**
fresh install

@MorrisJobke This is an additional issue than the partial translated activity notification mails in https://github.com/owncloud/activity/issues/355 which should be fixed in 8.1.2

00003575

### PR DESCRIPTION
**Pull request by MorrisJobke**
- fixes owncloud/activity#373

cc @LukasReschke @nickvergessen @PVince81

### CODE REVIEW

**Comment by MorrisJobke** [comment]
@ospringer Can you test this? This should save the detected language on login. If you login via webdav first, nothing is saved to DB.

File: lib/private/l10n.php [hunk 1]
```diff
@@ -463,13 +466,18 @@ public static function findLanguage($app = null) {
...
+		$default_language = $config->getSystemValue('default_language', false);
 
 		if($default_language !== false) {
 			return $default_language;
 		}
 
-		return self::setLanguageFromRequest($app);
+		$lang = self::setLanguageFromRequest($app);
+		if($userId && !$config->getUserValue($userId, 'core', 'lang')) {
+			$config->setUserValue($userId, 'core', 'lang', $lang);
```
Comment (DeepDiver1975) [inline lib/private/l10n.php]: I would not store the language as set by the request from the client as user default.
The request can com from a mobile client as well as a desktop client - this is unrelated to the user setting in the browser.

Comment (DeepDiver1975) [inline lib/private/l10n.php]: I'll talk with @nickvergessen in person here ....

Comment (DeepDiver1975) [inline lib/private/l10n.php]: :+1:

**Comment by MorrisJobke** [comment]
Oracle come on!

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Comment by MorrisJobke** [comment]
@karlitschek Can we backport this to stable8.1?

**Comment by karlitschek** [comment]
@MorrisJobke Backport make sense!

**Comment by OSpringer** [comment]
@MorrisJobke Thanks! I just applied the patch and quickly tested it. The notification mail was sent (partially) translated, without having manually selected a language. Great job!

**Comment by MorrisJobke** [comment]
stable8.1 https://github.com/owncloud/core/pull/18572

**Comment by nickvergessen** [comment]
@OSpringer the partly translated will be fixed in the next update as well, the reason was the files app used activity information incorrectly (and ignored the given language):
https://github.com/owncloud/core/pull/18171

**Pull request merged**